controllers/compaction/reconciler.go

```diff
@@ -384,8 +389,25 @@ func getCompactionJobVolumes(etcd *druidv1alpha1.Etcd, logger logr.Logger) []v1.
 		logger.Error(err, "Storage provider is not recognized. Compaction job will fail as no storage could be configured", "namespace", etcd.Namespace, "name", etcd.Name)
 		return vs
 	}
+	switch provider {
+	case "Local":
+		hostPath, err := utils.GetHostMountPathFromSecretRef(ctx, cl, logger, storeValues, etcd.Namespace)
+		if err != nil {
+			logger.Error(err, "Host mount path could not be detremined from provided secrets", "namespace", etcd.Namespace, "name", etcd.Name)
+			return vs
```

Comment:
If there is an error in getting host mount path, then you should return the error to the caller of `getCompactionJobVolumes()`, instead of returning `vs` without the correct volumes. You will need to enhance `getCompactionJobVolumes()` to also return an error, which is correctly handled in `createCompactionJob()`.

Reply:
I think it is better to not handle the error separately. I am raising the error in logger so the error will be visible in container log. Apart from that, if the mount path is not determined and an empty `vs` is returned, the compaction job will also raise another error of not able to recognise the volume.  I think, we are good here.

Reply:
Errors always need to be well-handled. Druid should make no assumption about how backup-restore would handle the misconfiguration. If druid is unable to get the volume path, then it should mandatorily return error, rather than knowingly return wrong volumes. Logging it is not sufficient, since that requires a human operator to dig through the logs to figure out what went wrong.

Reply:
See, you can see throughout the same file that errors are handled the same way as in this function. I would not like to change the behaviour for this particular case. If a compaction job fails, the operator needs to check logs anyway. If you have another suggestion to handle this error in better way, we can work that in further enhancement but not in this PR

Reply:
> you can see throughout the same file that errors are handled the same way

As a maintainer of the project, it would be advisable to fix the behaviour for the other error handling as well, rather than repeating previous mistakes again. If there are other instances in this file where errors are handled in a wrong way, I would actually advise to fix those as well.

> If a compaction job fails, the operator needs to check logs anyway

With well-defined and well-exposed error codes, an operator has no necessity to go through the logs of any component. As you already know, we are heading towards this direction for druid, and the first step is to ensure that errors are atleast returned correctly to the top level caller of functions.

> If you have another suggestion to handle this error in better way

I have already suggested a way to correctly handle the error, in my [initial comment] (https://github.com/gardener/etcd-druid/pull/682#discussion_r1341300449). It's a simple enough change. I would prefer if PRs don't introduce technical debt that needs to be handled later. As a maintainer, you would also realise the pain and effort it takes to correct past mistakes in code bases. Please ensure that this PR has correct error handling. For other instances of mishandling of errors, please raise an issue so that someone else might pick it up as a good first issue.

Reply:
I considered your suggestion. But I won't apply your suggestion for error handling in this PR. Also, the suggestion looks not ideal to me.

I also want to mention that error is handled in current design as well. If `utils.GetHostMountPathFromSecretRef` returns an error, compaction controller will print the error in its log. But the controller will also allow compaction jobs to be created with no volume. So, the compaction job will eventually fail after creation and post an error in it's description. It will be helpful this way because an operator, who looks into compaction dashboard, will easily see that the compaction jobs are failing and take action on that. Otherwise, if the compaction jobs were not created at all due to error in `utils.GetHostMountPathFromSecretRef`, an operator would not know the problem easily as he needs to find the log in etcd-druid in that case.

Considering the above-mentioned reasoning, I will not implement any other way of error handling in this PR. Thank you for your review.

Reply:
I don't agree with this reasoning. It's rudimentary to properly handle errors in the code. In this case, if the function encounters an error, it becomes its duty to pass that error back to the caller of the function to decide what action to take based on that information. But passing back a nil error and a wrong return value, ie, incomplete `vs` (volumes) for the compaction job makes no sense to me. Between catching the error within druid itself, and allowing it to deploy a compaction job (which takes up time and resources) and *then* catching error, I don't think there's any question that the former is preferable. I'm not sure why you still don't see that.

Reply:
@abdasgupta as agreed, can you please update the details of the out-of-band discussion that we had couple of days ago?

Reply:
In an out-of-band discussion with @shreyas-s-rao , I agreed to make the changes for stricter error handling in Druid itself. Druid won't create a compaction job if it identifies an error in the job spec. Though I still like Druid to be just a provisioner of compaction job and the flexibility of compaction job to run independently with minimum supervision of Druid, but our future design changes are intended to give more control on Druid for the objects it spawn. So, I agreed to change this PR to take care of error handling for volume mounts and for other scenarios.

Reply:
Changes are done